src/sciline/reporter.py

```diff
@@ -0,0 +1,256 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2025 Scipp contributors (https://github.com/scipp)
+
+from __future__ import annotations
+
+import functools
+from abc import ABC, abstractmethod
+from collections.abc import Callable, Hashable, Iterable
+from types import TracebackType
+from typing import Any
+
+from ._provider import Provider
+from ._utils import provider_name
+
+
+class Reporter(ABC):
+    """Base class for progress reporters of computations."""
+
+    def __init__(self) -> None:
+        self._n_steps = 0
+        self._provider_id = -1
+
+    def run_computation(self, providers: Iterable[Provider]) -> Reporter:
+        """Run a single computation with progress reporting.
+
+        A ``Reporter`` instance must not be used for multiple computations concurrently!
+
+        Use this method in a ``with`` statement:
+
+        .. code-block:: python
+
+            with reporter.run_computation(providers):
+                # Perform computation
+
+        This method exists mainly to provide arguments when entering a reporting
+        context, implementers should mainly override ``__enter__``.
+        """
+        self._n_steps = sum(1 for provider in providers if provider.kind == "function")
+        return self
```

Comment:
I'm a bit confused here. This is reporting some sort of progress, not exactly the estimated time to completion right? Different providers could take widly different times right?

Reply:
Yes, not the estimated time, see https://github.com/scipp/sciline/blob/015e7f275a1153d9ff368889db03251d169fb6e8/src/sciline/reporter.py#L119

The output looks like this (when the computation is completed):
```
Computing ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7/7 0:00:01
```